vertx-auth-ldap/src/main/java/io/vertx/ext/auth/ldap/impl/LdapAuthenticationImpl.java

```diff
@@ -41,68 +47,115 @@ public class LdapAuthenticationImpl implements LdapAuthentication {
+  /**
+   * Gets an authenticated user from an LDAP source.
+   *
+   * @param principal   The user principal that's authenticated
+   * @param credentials The username and password of the user
+   * @return A user that's been authenticated against the LDAP source
+   */
+  private Future<User> getUser(String principal, UsernamePasswordCredentials credentials) {
     Hashtable<String, Object> environment = new Hashtable<>();
-    // set the initial cntext factory
+    Promise<User> promise = ((VertxInternal) vertx).promise();
+
     environment.put(Context.INITIAL_CONTEXT_FACTORY, "com.sun.jndi.ldap.LdapCtxFactory");
-    // set the url
     environment.put(Context.PROVIDER_URL, authenticationOptions.getUrl());
 
     if (principal != null) {
       environment.put(Context.SECURITY_PRINCIPAL, principal);
     }
-    if (credential != null) {
-      environment.put(Context.SECURITY_CREDENTIALS, credential);
+
+    if (credentials != null) {
+      environment.put(Context.SECURITY_CREDENTIALS, credentials.getPassword());
     }
-    if (authenticationOptions.getAuthenticationMechanism() == null && (principal != null || credential != null)) {
+
+    if (authenticationOptions.getAuthenticationMechanism() == null && (principal != null || credentials != null)) {
       environment.put(Context.SECURITY_AUTHENTICATION, SIMPLE_AUTHENTICATION_MECHANISM);
     }
-    // referral
-    environment.put(Context.REFERRAL,
-      authenticationOptions.getReferral() == null ? FOLLOW_REFERRAL : authenticationOptions.getReferral());
 
-    Promise<LdapContext> promise = ((VertxInternal) vertx).promise();
+    environment.put(Context.REFERRAL,
+        authenticationOptions.getReferral() == null ? FOLLOW_REFERRAL : authenticationOptions.getReferral());
 
     vertx.executeBlocking(blockingResult -> {
       try {
         LdapContext context = new InitialLdapContext(environment, null);
-        blockingResult.complete(context);
+        User user = User.fromName(credentials.getUsername());
+
+        // Authentication method metadata
+        user.principal().put("amr", Collections.singletonList("pwd"));
+
+        // If filter query is set, we want to store additional user metadata which can
+        // be used later to authorize
+        if (authenticationOptions.getFilterQuery() != null) {
+          String filter = getFilterQuery(credentials.getUsername());
+          SearchControls searchControls = getSearchControls();
+          NamingEnumeration<SearchResult> searchResults = context.search(principal, filter, searchControls);
+
+          if (searchResults.hasMore()) {
+            SearchResult result = searchResults.next();
+            NamingEnumeration<? extends Attribute> attributes = result.getAttributes().getAll();
+            JsonObject metadata = user.principal();
+
+            // If attributes are returned, store them in our User object
+            while (attributes.hasMore()) {
```

Comment:
Yes, the intention was that someone would construct a filter that returns only one, but perhaps it should check that assumption to make sure there isn't more than one returned -- I'll add that.

Reply:
If the current impl works for our purposes, another option would be to scope this PR to just that. It could always be made more general purpose later (which seems like a lot of work, tbh).

Reply:
On second thought, I'll save this for when we either prepare for upstream or fork for our own purposes.

Reply:
I just added a log warning to indicate there might be a problem. Yes, there is definitely future work for making this something we'd submit back upstream or something we'd maintain.